### PR TITLE
feat: use cache kv manager for SchemaMetadataManager

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Add cache layer for SchemaMetadataManager

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Review by MichaelScofield — APPROVED** [review]
Lack of prefetching the database ttl option, are you gonna do it in next PR?

File: src/datanode/src/heartbeat.rs [hunk 1]
```diff
@@ -359,3 +366,48 @@ impl HeartbeatTask {
...
+            .await?;
+        Ok(HandleControl::Done)
+    }
+}
+
+impl InvalidateSchemaCacheHandler {
+    pub fn new(cached_kv_backend: Arc<CachedKvBackend>) -> Self {
+        Self { cached_kv_backend }
+    }
+}
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/datanode/src/heartbeat.rs:413]: By the way, we can refactor it in the future and move it to the `common/meta/handler` dir.

[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/datanode/src/heartbeat.rs]: fixed in 0876f6297d

**Comment by v0y4g3r** [comment]
> Lack of prefetching the database ttl option, are you gonna do it in next PR?

Here: https://github.com/GreptimeTeam/greptimedb/pull/5053/files#diff-c8c39874ac506c24bf14a603286e832fd039cdc068f8abec26d8e4061c255f32R64-R66

File: src/catalog/src/kvbackend/client.rs [hunk 2]
```diff
@@ -112,19 +110,29 @@ pub type CacheBackend = Cache<Vec<u8>, KeyValue>;
...
     version: AtomicUsize,
 }
 
-impl TxnService for CachedMetaKvBackend {
+#[async_trait::async_trait]
+impl TxnService for CachedKvBackend {
     type Error = Error;
+
+    async fn txn(&self, txn: Txn) -> std::result::Result<TxnResponse, Self::Error> {
+        //todo(hl): txn of CachedKvBackend simply pass through to inner backend without invalidating caches.
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/catalog/src/kvbackend/client.rs]: > 

fixed 2803a7f65d

**Review by fengjiachun — APPROVED** [review]

File: src/datanode/src/heartbeat/handler/cache_invalidator.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+        for cache in caches {
+            let CacheIdent::SchemaName(schema_name) = cache else {
+                continue;
+            };
+            let key: SchemaNameKey = (&schema_name).into();
+            let key_bytes = key.to_bytes();
+            // invalidate and refill cache
+            self.cached_kv_backend.invalidate_key(&key_bytes).await;
+            let _ = self.cached_kv_backend.get(&key_bytes).await?;
```
Comment (MichaelScofield) [inline src/datanode/src/heartbeat/handler/cache_invalidator.rs:66]: Seems it might block the heartbeat channel in the datanode?

**Comment by MichaelScofield** [comment]
> > Lack of prefetching the database ttl option, are you gonna do it in next PR?
> 
> Here: https://github.com/GreptimeTeam/greptimedb/pull/5053/files#diff-c8c39874ac506c24bf14a603286e832fd039cdc068f8abec26d8e4061c255f32R64-R66

@v0y4g3r do we need to prefetch the database options when opening the region in the datanode?

**Comment by v0y4g3r** [comment]
> > > Lack of prefetching the database ttl option, are you gonna do it in next PR?
> > 
> > 
> > Here: https://github.com/GreptimeTeam/greptimedb/pull/5053/files#diff-c8c39874ac506c24bf14a603286e832fd039cdc068f8abec26d8e4061c255f32R64-R66
> 
> @v0y4g3r do we need to prefetch the database options when opening the region in the datanode?

I think it's ok for the first time?

File: src/datanode/src/heartbeat/handler/cache_invalidator.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/datanode/src/heartbeat/handler/cache_invalidator.rs:66]: > I think we do not need to refill the cache

We need to prefetch the database otions on altering.

**Comment by MichaelScofield** [comment]
> > > > Lack of prefetching the database ttl option, are you gonna do it in next PR?
> > > 
> > > 
> > > Here: https://github.com/GreptimeTeam/greptimedb/pull/5053/files#diff-c8c39874ac506c24bf14a603286e832fd039cdc068f8abec26d8e4061c255f32R64-R66
> > 
> > 
> > @v0y4g3r do we need to prefetch the database options when opening the region in the datanode?
> 
> I think it's ok for the first time?

@v0y4g3r hmm... just some picky ones would complain about the first time performance. But let's not block this PR, I'll do it.

File: src/datanode/src/heartbeat/handler/cache_invalidator.rs (hunk 3 above)
Comment (fengjiachun) [inline src/datanode/src/heartbeat/handler/cache_invalidator.rs:66]:
```suggestion
```

File: src/datanode/src/heartbeat/handler/cache_invalidator.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            caches
+        );
+
+        for cache in caches {
+            let CacheIdent::SchemaName(schema_name) = cache else {
+                continue;
+            };
+            let key: SchemaNameKey = (&schema_name).into();
+            let key_bytes = key.to_bytes();
+            // invalidate and refill cache
```
Comment (fengjiachun) [inline src/datanode/src/heartbeat/handler/cache_invalidator.rs:64]:
```suggestion
            // invalidate cache
```

File: src/datanode/src/heartbeat/handler/cache_invalidator.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+use async_trait::async_trait;
+use catalog::kvbackend::CachedKvBackend;
+use common_meta::cache_invalidator::KvCacheInvalidator;
+use common_meta::heartbeat::handler::{
+    HandleControl, HeartbeatResponseHandler, HeartbeatResponseHandlerContext,
+};
+use common_meta::instruction::{CacheIdent, Instruction};
+use common_meta::key::schema_name::SchemaNameKey;
+use common_meta::key::MetadataKey;
+use common_meta::kv_backend::KvBackend;
```
Comment (fengjiachun) [inline src/datanode/src/heartbeat/handler/cache_invalidator.rs:28]:
```suggestion
```

**Pull request merged**